bson/vector.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (C) MongoDB, Inc. 2024-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package bson
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+)
+
+// These constants are vector data types.
+const (
+	Int8Vector      = 0x03
+	Float32Vector   = 0x27
+	PackedBitVector = 0x10
+)
+
+// These are vector conversion errors.
+var (
+	ErrNotVector              = errors.New("not a vector")
+	ErrInsufficientVectorData = errors.New("insufficient data")
+	ErrNonZeroVectorPadding   = errors.New("padding must be 0")
+	ErrVectorPaddingTooLarge  = errors.New("padding larger than 7")
+)
+
+// Vector represents a densely packed array of numbers.
+type Vector[T int8 | float32] struct {
+	Data []T
+}
+
+// BitVector represents a binary quantized (PACKED_BIT) vector of 0s and 1s
+// (bits). The Padding prescribes the number of bits to ignore in the final byte
+// of the Data. It should be 0 for an empty Data and always less than 8.
+type BitVector struct {
+	Padding uint8
+	Data    []byte
+}
+
+func newInt8Vector(b []byte) (Vector[int8], error) {
+	var v Vector[int8]
+	if len(b) == 0 {
+		return v, ErrInsufficientVectorData
+	}
+	if padding := b[0]; padding > 0 {
+		return v, ErrNonZeroVectorPadding
+	}
+	s := make([]int8, 0, len(b)-1)
+	for i := 1; i < len(b); i++ {
+		s = append(s, int8(b[i]))
+	}
+	v.Data = s
+	return v, nil
+}
+
+func newFloat32Vector(b []byte) (Vector[float32], error) {
+	var v Vector[float32]
+	if len(b) == 0 {
+		return v, ErrInsufficientVectorData
+	}
+	if padding := b[0]; padding > 0 {
+		return v, ErrNonZeroVectorPadding
+	}
+	l := (len(b) - 1) / 4
+	if l*4 != len(b)-1 {
+		return v, ErrInsufficientVectorData
+	}
+	s := make([]float32, 0, l)
+	for i := 1; i < len(b); i += 4 {
+		s = append(s, math.Float32frombits(binary.LittleEndian.Uint32(b[i:i+4])))
+	}
+	v.Data = s
+	return v, nil
+}
+
+func newBitVector(b []byte) (BitVector, error) {
+	var v BitVector
+	if len(b) == 0 {
+		return v, ErrInsufficientVectorData
+	}
+	padding := b[0]
+	if padding > 7 {
+		return v, ErrVectorPaddingTooLarge
+	}
+	if padding > 0 && len(b) == 1 {
+		return v, ErrNonZeroVectorPadding
+	}
+	v.Padding = padding
+	v.Data = b[1:]
+	return v, nil
+}
+
+// NewVectorFromBinary unpacks a BSON Binary into a Vector.
+func NewVectorFromBinary(b Binary) (interface{}, error) {
```

Comment:
Should we make this a method on the `Binary` type instead?

E.g.
```go
func (b Binary) Vector() Vector
func (b Binary) VectorErr() (Vector, error)
```

Reply:
IMO, the constructor(-alike) function is more conventional. Especially, we don't have to offer two flavors to deal with the error. Moreover, we may need to parse more types from the Binary in the future.

Reply:
Keeping the existing `NewVectorFromBinary` sounds good.